compress/zstd/zstd.go

```diff
@@ -0,0 +1,66 @@
+// Package zstd is an implementation of Compression interface using zstd format.
+package zstd
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/klauspost/compress/zstd"
+)
+
+// ZSTD implements Compression interface using gzip format.
+type ZSTD struct {
+	mutex  sync.Mutex
+	buffer bytes.Buffer
+	writer *zstd.Encoder
+	level  zstd.EncoderLevel
+}
+
+// New creates a new ZSTD instance.
+func New(level int) (*ZSTD, error) {
+	z := ZSTD{
+		level: zstd.EncoderLevel(level),
+	}
+
+	var err error
+	z.writer, err = zstd.NewWriter(&z.buffer, zstd.WithEncoderConcurrency(1), zstd.WithEncoderLevel(z.level))
+	if err != nil {
+		return nil, err
+	}
+
+	return &z, nil
+}
+
+// Compress will compress data using zstd format.
+// In case of any errors, it will return raw data without compression.
+func (z *ZSTD) Compress(data []byte) ([]byte, error) {
+	var err error
+
+	tmpBuffer := &z.buffer
+	tmpWriter := z.writer
+	if z.mutex.TryLock() {
+		defer z.mutex.Unlock()
+
+		tmpBuffer.Reset()
+		tmpWriter.Reset(&z.buffer)
+	} else {
+		// This should not happen since we only run a single thread.
+		// Good to have this logic here if we want to support concurrent processing.
+		tmpBuffer = new(bytes.Buffer)
+		tmpWriter, err = zstd.NewWriter(tmpBuffer, zstd.WithEncoderConcurrency(1), zstd.WithEncoderLevel(z.level))
+		if err != nil {
+			return data, err
+		}
+	}
+
+	_, err = tmpWriter.Write(data)
```

Comment:
noob question: what is the difference with the method that uses `io.Copy` and using `Write` [as suggested in the doc ](https://github.com/klauspost/compress/tree/master/zstd#usage)?

For small blocks, they encourage to use `EncodeAll`, do you know what difference it would make on our logs ?
Right now with max aggregation size of 150KB, we probably expect around 100 log events

Reply:
`io.Copy(enc, in)` copies from an `io.Reader`, deep down, it calls Write method of the `io.Writer`, same as we did, but it might also allocate some buffer. It is good if we deal with a Reader, like reading from a file of HTTP response of a stream. Here we have `[]byte` (the result of `proto.Marshal(...)`) and no Reader.

`Write` will encode on blocks, so it splits the data into several blocks and encodes them, `EncodeAll()` skips some checks and chunks of the data. The size of blocks depends on the compression level or if we overwrite it.
By default, for level 1: 64KB, for other levels: 128KB.

After compression (reduce log size by n folds), I thought we could increase max aggregation size n times.
For example, let's say our compression ratio is 5; then we can set the max aggregation size to 750KB and end up with ~150KB Kinesis events; otherwise, we will send 30KB events.

Reply:
Thanks for the clarification !

So EncodeAll could be potentially faster on small blocks ( smaller than the default block size ) ?

>After compression (reduce log size by n folds), I thought we could increase max aggregation size n times.
For example, let's say our compression ratio is 5; then we can set the max aggregation size to 750KB and end up with ~150KB Kinesis events; otherwise, we will send 30KB events.

Yes. This is exactly what we did inside the logging pipeline.

Reply:
> So EncodeAll could be potentially faster on small blocks ( smaller than the default block size ) ?

Yeah, slightly faster on smaller than the default block sizes.
Also, the compression ratio will be slightly different if your data size is bigger than the batch size (I haven't gone deep into the algorithms and codes, I might be wrong).
e.g. batch size 64KB, and you want to compress 70KB -> Write will compress the first 64KB, and then the following 6KB, while EncodeAll will compress the whole 70KB together. So I think it will end up with a smaller compressed size.